### PR TITLE
Don't throw an exception when a field is missing

### LINKED ISSUE #17202: Utility command --check-yaml crashes
When I make maps I always run this utility command on the newest bleed version. When I tried it, it crashed with this message. That command does not crash for other RA  versions (release and playtest) on my map
```
OpenRA.Utility(1,1): Error: Missing Type: EditorActionManagerInfo
OpenRA.Utility(1,1): Error: Failed with exception: System.AggregateException: One or more errors occurred. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at OpenRA.FieldLoader.Load (System.Object self, OpenRA.MiniYaml my) [0x00000] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.ActorInfo.LoadTraitInfo (OpenRA.ObjectCreator creator, System.String traitName, OpenRA.MiniYaml my) [0x00042] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.ActorInfo..ctor (OpenRA.ObjectCreator creator, System.String name, OpenRA.MiniYaml node) [0x0004b] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Ruleset+<>c__DisplayClass12_0.<LoadDefaults>b__1 (OpenRA.MiniYamlNode k) [0x00016] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Exts.ToDictionaryWithConflictLog[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector, System.String debugName, System.Func`2[T,TResult] logKey, System.Func`2[T,TResult] logValue) [0x0008e] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Ruleset.MergeOrDefault[T] (System.String name, OpenRA.FileSystem.IReadOnlyFileSystem fileSystem, System.Collections.Generic.IEnumerable`1[T] files, OpenRA.MiniYaml additional, OpenRA.IReadOnlyDictionary`2[TKey,TValue] defaults, System.Func`2[T,TResult] makeObject, System.Func`2[T,TResult] filterNode) [0x0006e] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Ruleset+<>c__DisplayClass12_0.<LoadDefaults>b__0 () [0x00000] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
  at System.Threading.Tasks.Task.Execute () [0x00010] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout) [0x00000] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
  at OpenRA.Ruleset.LoadDefaults (OpenRA.ModData modData) [0x00075] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.ModData.<.ctor>b__30_3 () [0x00000] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at System.Lazy`1[T].CreateValue () [0x00060] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Lazy`1[T].get_Value () [0x00020] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
  at OpenRA.ModData.get_DefaultRules () [0x00000] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Ruleset.LoadDefaultsForTileSet (OpenRA.ModData modData, System.String tileSet) [0x00000] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Map.PostInit () [0x00078] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Map..ctor (OpenRA.ModData modData, OpenRA.FileSystem.IReadOnlyPackage package) [0x0039b] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Mods.Common.UtilityCommands.CheckYaml.OpenRA.IUtilityCommand.Run (OpenRA.Utility utility, System.String[] args) [0x00165] in <c20bc88f1d7b4a3bb9778af5f76d4e82>:0 
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at OpenRA.FieldLoader.Load (System.Object self, OpenRA.MiniYaml my) [0x00000] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.ActorInfo.LoadTraitInfo (OpenRA.ObjectCreator creator, System.String traitName, OpenRA.MiniYaml my) [0x00042] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.ActorInfo..ctor (OpenRA.ObjectCreator creator, System.String name, OpenRA.MiniYaml node) [0x0004b] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Ruleset+<>c__DisplayClass12_0.<LoadDefaults>b__1 (OpenRA.MiniYamlNode k) [0x00016] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Exts.ToDictionaryWithConflictLog[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector, System.String debugName, System.Func`2[T,TResult] logKey, System.Func`2[T,TResult] logValue) [0x0008e] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Ruleset.MergeOrDefault[T] (System.String name, OpenRA.FileSystem.IReadOnlyFileSystem fileSystem, System.Collections.Generic.IEnumerable`1[T] files, OpenRA.MiniYaml additional, OpenRA.IReadOnlyDictionary`2[TKey,TValue] defaults, System.Func`2[T,TResult] makeObject, System.Func`2[T,TResult] filterNode) [0x0006e] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at OpenRA.Ruleset+<>c__DisplayClass12_0.<LoadDefaults>b__0 () [0x00000] in <4f4232c09ba042bc87d8d1ed0743357b>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
  at System.Threading.Tasks.Task.Execute () [0x00010] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 <---
```

### PR DESCRIPTION
**Pull request by abcdefg30**
Closes #17202.

Testcase: Add `JunkTrait:` to the world actor in `./mods/ra/maps/allies-01/rules.yaml`.
On bleed:
```
> OpenRA.Utility.exe ra --check-yaml ./mods/ra/maps/allies-01
OpenRA.Utility(1,1): Error: Missing Type: JunkTraitInfo
Testing map: 01:   In the thick of it
OpenRA.Utility(1,1): Error: System.AggregateException: Mindestens ein Fehler ist aufgetreten. ---> System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei System.Object.GetType()
   bei OpenRA.FieldLoader.Load(Object self, MiniYaml my)
   bei OpenRA.ActorInfo.LoadTraitInfo(ObjectCreator creator, String traitName, MiniYaml my)
   bei OpenRA.ActorInfo..ctor(ObjectCreator creator, String name, MiniYaml node)
   bei OpenRA.Ruleset.<>c__DisplayClass14_0.<Load>b__1(MiniYamlNode k)
   bei OpenRA.Exts.ToDictionaryWithConflictLog[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, String debugName, Func`2 logKey, Func`2 logValue)
   bei OpenRA.Ruleset.MergeOrDefault[T](String name, IReadOnlyFileSystem fileSystem, IEnumerable`1 files, MiniYaml additional, IReadOnlyDictionary`2 defaults, Func`2 makeObject, Func`2 filterNode)
   bei OpenRA.Ruleset.<>c__DisplayClass14_0.<Load>b__0()
   bei System.Threading.Tasks.Task.Execute()
   --- Ende der internen Ausnahmestapelüberwachung ---
   bei System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   bei System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   bei OpenRA.Ruleset.Load(ModData modData, IReadOnlyFileSystem fileSystem, String tileSet, MiniYaml mapRules, MiniYaml mapWeapons, MiniYaml mapVoices, MiniYaml mapNotifications, MiniYaml mapMusic, MiniYaml mapSequences, MiniYaml mapModelSequences)
   bei OpenRA.Map.PostInit()
---> (Interne Ausnahme #0) System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei System.Object.GetType()
   bei OpenRA.FieldLoader.Load(Object self, MiniYaml my)
   bei OpenRA.ActorInfo.LoadTraitInfo(ObjectCreator creator, String traitName, MiniYaml my)
   bei OpenRA.ActorInfo..ctor(ObjectCreator creator, String name, MiniYaml node)
   bei OpenRA.Ruleset.<>c__DisplayClass14_0.<Load>b__1(MiniYamlNode k)
   bei OpenRA.Exts.ToDictionaryWithConflictLog[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, String debugName, Func`2 logKey, Func`2 logValue)
   bei OpenRA.Ruleset.MergeOrDefault[T](String name, IReadOnlyFileSystem fileSystem, IEnumerable`1 files, MiniYaml additional, IReadOnlyDictionary`2 defaults, Func`2 makeObject, Func`2 filterNode)
   bei OpenRA.Ruleset.<>c__DisplayClass14_0.<Load>b__0()
   bei System.Threading.Tasks.Task.Execute()<---
```
This PR:
```
> OpenRA.Utility.exe ra --check-yaml ./mods/ra/maps/allies-01
OpenRA.Utility(1,1): Error: Missing Type: JunkTraitInfo
Testing map: 01:   In the thick of it
Errors: 1
```

### CODE REVIEW

**Comment by GraionDilach** [comment]
This shouldn't be included in the game, only within Utility at best.

If a trait ends up missing, that can easily be hidden within the game, and this also goes against the mindset which ended up enforcing typoed junk trait tags being logged.

**Comment by abcdefg30** [comment]
> This shouldn't be included in the game, only within Utility at best.

This is exactly what is happening here. Maybe the title is a bit misleading but if you read the actual code you'll see that the utility overwrites `MissingTypeAction` not to throw but only log an error instead. This PR is fixing a crash that happens as result of only logging instead of crashing. The normal game will still crash for junk values.

**Review by Mailaender — DISMISSED** [review]

**Comment by pchote** [comment]
As discussed in IRC, I have some reservations about the way this is implemented and think we should spend some more time looking at the problem before merging this.

**Comment by abcdefg30** [comment]
I am not sure what our alternative here is. As I told you already on IRC, we can't simply try-catch the error during the lint run (or crash without catching), since that would abort the rule parsing. (This is exactly what @chrisf wanted to prevent when he introduced this change way back in https://github.com/OpenRA/OpenRA/commit/5d1ee145e1c8dc331b6bedc529a5e96c3b73f730!)

However, since both you and @GraionDilach completely misunderstood the code and changes here after giving them a quick look, I'd suggest making what is going on in `ObjectCreator` a bit more obvious:
Default MissingTypeAction to null and change
https://github.com/OpenRA/OpenRA/blob/cadbd0d9ab3b03fa4c4e0396ba6ceb58de7d7e5c/OpenRA.Game/ObjectCreator.cs#L90
to something like
```
// HACK: The linter does not want to crash but only print an error instead
if (MissingTypeAction != null)
	MissingTypeAction(className);
else
	throw new InvalidOperationException("Cannot locate type: {0}".F(className));
```

That way it becomes more obvious that *only* the linter is omitting the exception. The added null checks will have to stay then though.

**Comment by pchote** [comment]
Sounds good to me. Can we please also have comment(s?) in `ActorInfo` clarifying that `CreateObject` will throw by default, or return null if `MissingTypeAction` is set from the linter?

**Comment by abcdefg30** [comment]
Updated with comments.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
There have been updates sine @Mailaender approved, so I think this still needs another approval.

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b48f7da89a206fd90bc20e891a34569e2b523d0c...1a6b82319183d8237d2f36b94fb0ebf85e1583d7

**Pull request merged**
